command/agent/dns.go

```diff
@@ -33,6 +33,37 @@ type DNSServer struct {
 	logger       *log.Logger
 }
 
+// Shutdown stops the DNS Servers
+func (d *DNSServer) Shutdown() error {
+	// Async stop the DNS Servers, handle a potential error
+	errCh := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServer.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping udp server: %v", err)
+			errCh <- fmt.Errorf("dns udp teardown failed: %v", err)
+		}
+	}(d)
+
+	errChTCP := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServerTCP.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping tcp server: %v", err)
+			errChTCP <- fmt.Errorf("dns tcp teardown failed: %v", err)
+		}
+	}(d)
+
+	// Wait for either the check, listen error, or timeout
```

Comment:
Hm, there is no exit path from this select other than error or timeout. We could introduce a third channel to signal completion, but do the DNS servers take more than a few milliseconds to shut down? It might be simpler to just do the shutdown in sequence.